log/stat/stat.go

```diff
@@ -0,0 +1,107 @@
+package stat
+
+import (
+	"bytes"
+	"fmt"
+	"sync"
+	"text/tabwriter"
+
+	"github.com/peak/s5cmd/strutil"
+)
+
+const (
+	totalCount = iota
+	succCount
+)
+
+var (
+	enabled bool
+	stats   statistics
+)
+
+type statistics [2]syncMap
+
+// InitStat initializes collecting program statistics.
+func InitStat() {
+	enabled = true
+	for i := range stats {
+		stats[i] = syncMap{
+			Mutex:       sync.Mutex{},
+			mapStrInt64: map[string]int64{},
+		}
+	}
+}
+
+// syncMap is a statically typed and synchronized map.
+type syncMap struct {
+	sync.Mutex
+	mapStrInt64 map[string]int64
+}
+
+func (s *syncMap) add(key string, val int64) {
+	s.Lock()
+	defer s.Unlock()
+
+	s.mapStrInt64[key] += val
+}
+
+// Stat is for storing a particular statistics.
+type Stat struct {
+	Visited    string `json:"visited"`
+	SuccVisits int64  `json:"success visits"`
+	ErrVisits  int64  `json:"error visits"`
+}
+
+// Collect collects function execution data.
+func Collect(path string, err *error) func() {
+	return func() {
+		if !enabled {
+			return
+		}
+		if err == nil || *err == nil {
+			stats[succCount].add(path, 1)
+		}
+		stats[totalCount].add(path, 1)
+	}
+}
+
+// Stats implements log.Message interface.
+type Stats []Stat
+
+func (s Stats) String() string {
+	b := bytes.Buffer{}
+
+	w := tabwriter.NewWriter(&b, 5, 0, 5, ' ', tabwriter.AlignRight)
+
+	fmt.Fprintf(w, "\n%s\t%s\t%s\t%s\t\n", "Operation", "Total", "Error", "Success")
+	fmt.Fprintf(w, "%s\t%s\t%s\t%s\t\n", "=========", "=====", "=====", "=======")
+	for _, stat := range s {
```

Comment:
Can you check the size of the stats before preparing a tabular format (if it is empty, we can return empty string)? If there is no stat present, it prints out the following output:

```
./s5cmd --stat
NAME:
   s5cmd - Blazing fast S3 and local filesystem execution tool

USAGE:
   s5cmd [global options] command [command options] [arguments...]

COMMANDS:
   ls       list buckets and objects
   cp       copy objects
   rm       remove objects
   mv       move/rename objects
   mb       make bucket
   du       show object size usage
   cat      print remote object's contents to stdout
   run      run commands in batch
   version  print version
   help, h  Shows a list of commands or help for one command

GLOBAL OPTIONS:
   --json                         enable JSON formatted output (default: false)
   --numworkers value             number of workers execute operation on each object (default: 256)
   --retry-count value, -r value  number of times that a request will be retried for failures (default: 10)
   --endpoint-url value           override default S3 host for custom services
   --no-verify-ssl                disable SSL certificate verification (default: false)
   --log value                    log level: (debug, info, error) (default: "info")
   --install-completion           install completion for your shell (default: false)
   --stat                         collect statistics of program execution and display it at the end (default: false)
   --help, -h                     show help (default: false)

     Operation     Total     Error     Success
     =========     =====     =====     =======
```

Reply:
thanks. done.